src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs

```diff
@@ -718,5 +719,93 @@ private async ValueTask<T> CatchAndRethrowExceptionAsync<T>(Func<Task<T>> taskFu
                 throw exception;
             }
         }
+
+        public async ValueTask<IReadOnlyDictionary<Tuple<NuGetFramework, string>, IReadOnlyList<IPackageReferenceContextInfo>>> GetTransitivePackageOriginAsync(PackageIdentity transitivePackage, string projectId, CancellationToken ct)
+        {
+            /** Pseudocode
+            1. Get project restore graph 
+
+            2. Filter by packages
+
+            3. Foreach direct dependency d:
+                do DFS to look for transitive dependency
+
+                if found:
+                  Add to list
+
+            4. Return list
+            */
+            ct.ThrowIfCancellationRequested();
+
+            var singleProjectId = new[] { projectId };
+            var projectsList = await GetProjectsAsync(singleProjectId, ct);
+            var project = projectsList.FirstOrDefault();
+            var packageOrigins = new Dictionary<Tuple<NuGetFramework, string>, IReadOnlyList<IPackageReferenceContextInfo>>();
+
+            if (project != default && project is PackageReferenceProject prProject)
+            {
+                IList<ProjectModel.LockFileTarget> fxGraphList = await prProject.GetFullRestoreGraphAsync(ct);
+
+                if (fxGraphList != null)
+                {
+                    var pkgs = await GetInstalledAndTransitivePackagesAsync(singleProjectId, ct);
+
+                    var visited = new HashSet<object>();
+
+                    foreach (var targetFxGraph in fxGraphList)
+                    {
+                        var key = Tuple.Create(targetFxGraph.TargetFramework, targetFxGraph.RuntimeIdentifier);
+                        var list = new List<IPackageReferenceContextInfo>();
+
+                        foreach (var directPkg in pkgs.InstalledPackages) // are InstalledPackages direct dependencies only? Yes!
+                        {
+                            visited.Clear();
+                            var found = FindTransitive(directPkg.Identity, transitivePackage, targetFxGraph, visited);
+                            if (found)
+                            {
+                                list.Add(directPkg);
+                            }
+                        }
+
+                        if (list.Any())
+                        {
+                            packageOrigins[key] = list;
+                        }
+                    }
+                }
+            }
+
+            return packageOrigins;
+        }
+
+        private bool FindTransitive(PackageIdentity current, PackageIdentity transitivePackage, ProjectModel.LockFileTarget graph, HashSet<object> visited)
+        {
+            if (current.Equals(transitivePackage))
+            {
+                return true;
+            }
+
+            var node = graph
+                .Libraries
+                .Where(x => x.Name == current.Id && x.Version.Equals(current.Version))
```

Comment:
I think it's best to unwrap this Linq to prevent a lot of extra allocations since this is on a code path with a double for-loop for targets and direct packages.